### ISSUE TITLE
[Bug] Tensor add performance issue

### ISSUE THREAD

**Issue by suenar** [hw-native-sys/pypto#1919]
### Component

Parser

### Description

Drring our tensor add test, the pypto performance is very slow while the torch-npu is fine.

Our test script:
```python
import pypto.language as pl
import torch
from pypto.runtime import RunConfig
import time 

@pl.jit
def tile_add(a: pl.Tensor, b: pl.Tensor, c: pl.Out[pl.Tensor]):
    with pl.incore():
        tile_a = pl.load(a, [0, 0], [128, 128])
        tile_b = pl.load(b, [0, 0], [128, 128])
        tile_c = pl.add(tile_a, tile_b)
        pl.store(tile_c, [0, 0], c)
    return c

ITER = 100
if __name__ == "__main__":
    #pypto test
    a = torch.randn(128, 128)
    b = torch.randn(128, 128)
    c = torch.zeros((128, 128), dtype=torch.float32)

    for i in range(10):
        tile_add(a, b, c)
    start = time.time()
    for i in range(ITER):
        tile_add(a, b, c)
    end = time.time()
    print("time:", end-start)

    #torch npu test
    x = torch.randn(128, 128)
    y = torch.randn(128, 128)

    for _ in range(10):
        z = (x.npu() + y.npu()).cpu()

    start = time.time()
    for i in range(ITER):
        z = (x.npu() + y.npu()).cpu()
    end = time.time()
    print("time:", end-start)
``` 

Out output:
time: 3.1880862712860107
time: 0.01300811767578125

Our test was run on chip Ascend910_9362

Our pypto related environment:
cann 8.5.0
driver 25.5.2
torch_npu 2.8.0
PTOAS commit: f59e6cb7e67a251426801f5f0958e4214174f5f1
pto-isa commit: b9122ec586b5a7b4b686ca7498874a1c94b3573c

Thanks for the help!

### Steps to Reproduce

```markdown
Run the python script above.
```

### Expected Behavior

pypto reported time is close to torch-npu version.

### Actual Behavior

pypto reported time is much slower than the torch-npu version.

### Git Commit ID

1eec39f1c3efb2460d33be14d1759cdc884c8c6f

### NPU Kind

Ascend 910C

### Host Platform

Linux (x86_64)

**Comment by YunjiQin** [comment]
## Root cause: the add kernel is ~7.8 µs — the rest is pypto's per-launch host overhead + AICPU orchestration/scheduling

I profiled this on real hardware (`platform="a2a3"`). The actual on-device compute is tiny; almost all of the wall time is fixed per-launch overhead that does not shrink for a trivial op. **This is not a codegen/kernel problem.**

### 1. Per-iter comparison (on-board, median)

| stage | direct call | worker reuse | benchmark (register-once) |
| --- | ---: | ---: | ---: |
| **per-iter (wall)** | **44.8 ms** | **14.1 ms** | **4.1 ms** |
| host — worker `init`/`register`/`close` | 29.1 ms | — | — |
| host — `compile_and_assemble` (per call) | 9.4 ms | 7.8 ms | — |
| host — other python (`_resolve`/`_coerce`/orch_args) | ~0 | 1.4 ms | — |
| `run_prepared` (one launch) | 6.5 ms | 4.9 ms | 4.1 ms |

Two per-call host taxes dominate and never amortize for a tiny op:
- **Worker rebuild (~29 ms/call):** with no active worker, every call constructs a device `Worker` + `init()` + `register()` + `close()` (`device_runner.execute_on_device`).
- **`compile_and_assemble` re-run (~8 ms/call):** `runner.execute_compiled` re-assembles the `ChipCallable` on *every* call (gcc toolchain probe + re-assembly from disk-cached binaries), bypassing the cached `CompiledProgram._chip_callable`.

Fix on the user side: use the register-once path (`pypto.runtime.benchmark(compiled, [a,b,c], rounds=N)`, or hold a `ChipWorker` open and dispatch a registered handle) → ~45 ms drops to ~4 ms. Runtime fix: cache the assembled `ChipCallable` in `execute_compiled` instead of re-running `compile_and_assemble` each call.

### 2. Where the ~4 ms floor goes (single launch, register-once)

Even with all host taxes removed, one launch is ~4 ms — and only **7.8 µs of it is the kernel**:

| stage (one launch) | time | what it is |
| --- | ---: | --- |
| `run_prepared` (total) | 4093 µs | one launch, the irreducible floor |
| ├ `bind.args` | 89 µs | host: bind call args |
| ├ `bind.prebuilt` | 1826 µs | host: C++ re-bind of the prebuilt program |
| ├ `runner_run` host (launch + sync) | 996 µs | host: kernel launch + wait |
| ├ `device_wall` — **AICPU orchestration** | 907 µs | on-device scheduler runtime (see below) |
| │  ├ `graph_build` | 591 µs | build the task graph on AICPU |
| │  ├ `preamble` | 282 µs | ring buffer / tensormap setup |
| │  ├ `post_orch` | 107 µs | teardown |
| │  └ `sched` + `orch` | 12 µs | task dispatch / schedule |
| ├ **actual kernel (AIV compute)** | **7.8 µs** | the only real work; sits inside `device_wall` |
| └ `validate` | 86 µs | host: result check |

### 3. Why torch-npu is fast: pypto carries an AICPU orchestration/scheduling runtime that torch-npu does not

This is the key difference. On every launch pypto's AICPU runtime **builds a task graph** (`graph_build` ~0.59 ms), **sets up rings/tensormap** (`preamble` ~0.28 ms), and **runs a task scheduler** to drive the kernel — the swimlane shows the full per-task pipeline `resolve → wire → submit → dispatch → complete` (each ~1–10 µs) feeding the single `tile_add` AIV task. That ~0.9 ms of on-device orchestration + scheduling is **fixed per launch, independent of op size**.

torch-npu has none of this: it dispatches a pre-registered aclnn op straight onto a persistent ACL stream — no task-graph build, no per-op scheduler. So for a 7.8 µs kernel its per-op overhead is far lower. pypto's scheduling model pays off when a program has many tasks (the orchestration amortizes), but for a single trivial op the scheduling overhead dominates the actual compute by ~100×.

### Note on the reproducer

The script uses the default `RunConfig`, whose platform is `a2a3sim` (**simulator**) — so it compares pypto-on-simulator against torch-on-real-NPU. All numbers above are on-board (`a2a3`) for a fair comparison; even there the fixed host + scheduling overhead dominates, so the conclusion stands.

**Comment by suenar** [comment]
Thanks for the help!
